### ISSUE TITLE
standardize async dbapi exceptions with access pattern

### ISSUE THREAD

**Issue by zzzeek** [sqlalchemy/sqlalchemy#8047]
### Discussed in https://github.com/sqlalchemy/sqlalchemy/discussions/8023

I think we should try to hit this in terms of all the async dialects all of which have to do what we are doing here, which would be that we make a common mixin / base for all the async-adapted DBAPIs.  for asyncpg it would be used here: https://github.com/sqlalchemy/sqlalchemy/blob/19f0566dc98cc1d088777a9f1e1e7e898d861c05/lib/sqlalchemy/dialects/postgresql/asyncpg.py#L760-L761

```py
    from ..exc import EmulatedDBAPIException

    class Error(EmulatedDBAPIException):
        pass
```

that way people can look at `EmulatedDBAPIException` to see the fields, which would be like:

```py
class EmulatedDBAPIException:
    """an exception class used to adapt non-DBAPI database driver exceptions
    into DBAPI exceptions"""

    message: str
    """string message indicating the error"""

    orig: Exception
    """the original driver-level exception object"""
```

that looks kind of boring so we can also add an accessor on https://github.com/sqlalchemy/sqlalchemy/blob/19f0566dc98cc1d088777a9f1e1e7e898d861c05/lib/sqlalchemy/exc.py#L525 called "driver_exception".  This would work similarly to how we have "dbapi_connection" / "driver_connection" on Connection.   

that way someone that knows asyncpg can receive the top level SQLAlchemy DBAPIException and just refer to .driver_exception to get all the asyncpg fields.

**Comment by zzzeek** [comment]
if someone wants to work on this sooner, it can be 2.0, but for now not an urgent need here

**Comment by bhperry** [comment]
This would be great. Currently in order to catch nowait lock errors from asyncpg I am doing string search for "LockNotAvailable" in the error message. Would much prefer to have access to the actual asyncpg exception.

**Comment by DeoLeung** [comment]
looking forward to get the origin exception as well
currently I can do `e.orig.pgcode` and match to the origin exception

**Comment by bhperry** [comment]
Is pgcode an int code?

**Comment by DeoLeung** [comment]
nope, but looking at the `asyncpg.exception`, it seems ok to do `.startswith('42')` to match to `SyntaxOrAcceessError`, same of other Exceptions :)

**Comment by jmehnle** [comment]
I believe I'm running into this issue. I'm running SQLAlchemy 2.0 on top of `asyncpg`, and when some database statement returns, e.g., a `not_null_violation` PostgreSQL error, SQLAlchemy raises a `sqlalchemy.exc.IntegrityError` exception. However, when I try to inspect that exception's `orig` attribute, what I see is an exception that, by its `__str__` value, pretends to be a `asyncpg.exceptions.NotNullViolationError` exception (as I would expect) but in reality is some obscure custom exception class with the following inheritance chain:
- `sqlalchemy.dialects.postgresql.asyncpg.AsyncAdapt_asyncpg_dbapi.IntegrityError`
- `sqlalchemy.dialects.postgresql.asyncpg.AsyncAdapt_asyncpg_dbapi.DatabaseError`
- `sqlalchemy.dialects.postgresql.asyncpg.AsyncAdapt_asyncpg_dbapi.Error`
- `Exception`
- `BaseException`

I think https://docs.sqlalchemy.org/en/20/core/exceptions.html#sqlalchemy.exc.DBAPIError makes an attempt at explaining this, but unfortunately that explanation went completely past my head.

SQLAlchemy's `IntegrityError` exception is too general as I need to distinguish between `not_null_violation`, `unique_violation`, and other types of PostgreSQL errors that map to `IntegrityError`. How can I inspect the original exception without resorting to string matching? What is the reason that `StatementError.orig` doesn't *actually* expose the original exception but rather some chimera version of it?

**Comment by DeoLeung** [comment]
```python
    except ProgrammingError as e:
      if e.orig.pgcode == UndefinedTableError.sqlstate:
```

hmmm, still string matching but a bit better workaround :)

**Comment by CaselIT** [comment]
> How can I inspect the original exception without resorting to string matching? What is the reason that StatementError.orig doesn't actually expose the original exception but rather some chimera version of it?

That was done to respect the pep249 exception hierarchy. The original exception is the cause of the custom pep249 excenption, since that's raised using `raise new_exc from asyncpg_exc`

**Comment by jmehnle** [comment]
Interesting. If I understood you correctly, then the way to get at the database driver's underlying exception is `e.__cause__`? Or `e.orig.__cause__`? Either way, it would be great if that was documented.

**Comment by zzzeek** [comment]
im looking to see if I can get the implementation proposed in for this, but i would note most of these comments seem to be lamenting issues that are intrinsic to the way DBAPIs are implemented, which is that yes, you need to catch the `IntegrityError` and do string matching; there are ways this kind of work can be automated and even ways that a nice 3rd party plugin could provide a deep standard exception hierarchy (building on [ExceptionContext](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.ExceptionContext)) but that's outside of current scope.


SQLAlchemy has to inspect string messages to determine "disconnect" exceptions across all backends, and I've had to do this in other projects across a wide range of exception types to correctly detect various conditions.      the fix here is not going to solve much of that.

**Comment by zzzeek** [comment]
> Interesting. If I understood you correctly, then the way to get at the database driver's underlying exception is `e.__cause__`? Or `e.orig.__cause__`? Either way, it would be great if that was documented.

@jmehnle im looking at the issue you describe and it seems very simply that the asyncpg dialect in SQLAlchemy is not actually saving the original asyncpg exception, and is throwing it away.  to be clear, I wrote this code, but this seems really surprising since I also wrote the code that gives us `e.orig` in the first place, so the idea of "attach the original exception" is not new to me, i dont know why i did it this way.   Definitely a bug, so when i do this exception hierarchy ill also make sure that orig_async_exception (name TBD) is definitely there.

**Comment by zzzeek** [comment]
it does look like `err.orig.__cause__` has that for the moment, so I might look to repurpose `__cause__` to provide that also here (that way whatever garbage collection issues might exist they aren't on sqlalchemy)

edit:  nah.   we store orig already, trying to ensure all throws are done so that `__cause__` lines up seems pretty shaky

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **main** branch:

**Improve asyncpg exception hierarchy and asyncio hierarchies overall** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6158

**Comment by jmehnle** [comment]
@zzzeek, thank you for fixing this!
